### DISCUSSION TITLE
[Suggestion] Incorporate Structured Data

### DISCUSSION THREAD

**Discussion by lat9** [zencart/zencart#5777] [Ideas]
As provided in @torvista's repository: https://github.com/torvista/zen-cart_Structured-Data

**Comment by torvista** [comment]
blimey, praise indeed!! It's not pretty...it's been evolved piecemeal to use whatever variables are already available at the time. But it works/is better than nothing.

**Comment by drbyte** [comment]
How "ready" is it?
How up-to-date is it? 
How does it compare with the other similar ones that it's derived from?

Who's going to keep it up-to-date with all the moving-target changes required by the 3rd-party-services which drive it?

**Comment by torvista** [comment]
> How "ready" is it?

As far as I am aware there are no unresolved issues with it.

> How up-to-date is it?
How does it compare with the other similar ones that it's derived from?

The pros will need to answer that.

**Comment by lankeeyankee** [comment]
this is the best version, no doubt. I use it, it's essential for every store owner, definitely needs to be in the core.

**Comment by lat9** [comment]
For insertion into the core, we could add a couple of notifiers to allow elements to be added or modified from the base.

**Comment by torvista** [comment]
It is long since I looked at the code but I was always unhappy at the bit-by-bit nature of how each piece of data was extracted from the environment or if not available, directly from the database.
Not sure if notifiers would help as no core files are harmed at the moment.
It's more the requirement of needing extra sql queries to get at info that has not been already created/is not available in the environment at that moment (according to superglobals).
Maybe some things could be created sooner, to be available for this mod to use, (but I'm not going to suggest that after the recent messagestack saga...although it did expose a bug)!

In the core?
yes:  the funcionality of this mod seems to be regarded as essential
no: the install is very simple and there is possible maintenance involved...
but very little was necessary in the last year. If a change is required and people don't implement it, so what.
yes: It's something that can go on the outside of the 157 tin as a new feature (bit of marketing really....although that is all the mod actually IS!!),  that requires little work to include.

**Comment by drbyte** [comment]
Is there a way to not stick hundreds of extra configuration entries into the database for it?
Do all of those settings need to be configurable from the admin in the browser? or are they set-it-and-forget-it enough that they could be a simple PHP array in a PHP file?

**Comment by torvista** [comment]
The balance between insulating users from editing code and admin bloat is not always clear. Most mod settings could be in a php file/are never changed.
Should the options for enable/disable still be in the admin? If so, that means a specific page for the mod anyway.

**Comment by zcadditions** [comment]
IMO,

This is an absolute must! The benefits of including structured data into the core (which it really doesn't touch) by far outweighs any admin (on/off) switch or including a complete admin interface (ie Tools).

**Comment by wiztechinc** [comment]
Obviously a must for some - I've never used it. Perhaps that shows my ignorance? It is simple for sure and does not touch core and not a big deal to incorporate but how much maintenance does it require? I think that should be the most important part of this conversation. Plus this is definitely a developer thing: my clients would take one look and run screaming for the hills. I used to develop modules for facebook and gave up since they continue to change things so regularly. If including something like google analytics isn't considered necessary or too high maintenance, I'm really unsure about a part of this being so intertwined with social media. I also hate the amount of source code this generates in the end. If someone wants to add all this in, than go for it. Making it part of the core? Not so much.

**Comment by zcadditions** [comment]
IMO

NOT including structured data is a huge injustice to a client and a huge mistake for any website owner.

Learn it, promote it and implement it!
https://yoast.com/structured-data-schema-ultimate-guide/

**Comment by lankeeyankee** [comment]
> IMO
> 
> NOT including structured data is a huge injustice to a client and a huge mistake for any website owner.
> 
> Learn it, promote it and implement it!
> https://yoast.com/structured-data-schema-ultimate-guide/

It's really a must-have for everyone in 2020, there is no situation where you would not want this.

**Comment by wiztechinc** [comment]
The question then becomes what other ecommerce projects or CMS projects have structured data in the core? WordPress definitely doesn't. I'm not quibbling with how necessary this is or isn't. Should it be in the core?

**Comment by barco57** [comment]
ok, I am torn on this one and have removed my thumbs up from the pro side of the argument. I spent some time talking to one of the participants on the opposite side, and then I went and did some research. 
First off let me say that I normally am on the side of shove all the things into the core as it will save me the trouble of having to figure out the sometimes complicated merges to get everything to work together and force plugin/template authors to account for those things in their work. Ok, back on the point cause I am about to get ranty and Chris has probably heard enough of that from me on Skype and on the phone.
In my research, I found that Wordpress, Prestashop, Magento, Open Cart, Cube Cart, and Shopify do not provide structured data markup in the core application. It is always either a plugin/marketplace app. There must be a good reason for this, from too much maintenance in an ever-changing landscape to not wanting to give up the revenue share of market place apps, but there is no clear answer that I have yet found in the short time I researched. I suspect the answer is closer to the ever-changing landscape which is the good Dr's concern. 
zcadditions points to a very good article on Structured data in general from the #1 SEO plugin author for Wordpress who is basically pointing out there is a lot to it, a lot to learn and ends with links to their plugin. So I am not sure it makes the argument for it being in the core. I do know that on the couple of Wordpress installs I look after the Yoast plugin updates all the time, sometimes improvement, sometimes bug fixes, and sometimes to deal with a change/addition Google, Facebook, etc made. With what I see as the number of changes to structured data over time held up against the ZC release cycle (is there a release cycle? lol) it would be difficult to keep up if it was in the core whereas as a plugin much easier.

**Comment by torvista** [comment]
While changes due to target platform requirements have been few and far between, there is always that likelihood, and this is guaranteed to happen the day after a ZC release.
I think that until there are admin/internal mechanisms for both the notification of a plugin update and a click-to-update function, it should stay as an external module, especially since it is an easy install.
Meanwhile, a peer review to add any relevant notifiers would not go amiss.

**Comment by scottcwilson** [comment]
I think I'd rather keep this as a plugin, since Google's rules about what is expected in structured data can and do change.

**Comment by torvista** [comment]
I think there is a case for having this functionality included in core as it is a basic necessity for an ecommerce solution and should be listed as part of what is in the ZC box to attract new users. New users compare functionality, not the underlying platform/code quality.
Whether it uses my code or others is another issue.
Things do change, so ideally there should be a notification and additionally an update mechanism.
But in that case the responsible store owner will receive notification from Google when things go awry.
In reality things change rarely, so I think the advertised advantage of having it in core outweighs the possible need for an update within the ZC releases. ZC is not intended to be install and forget so the occasional wake up for a user is not a deal-breaker, I don't think.

**Comment by neekfenwick** [comment]
Definitely think this would be good as a plugin. I don't know much about the 'new' (to me, anyway) plugin architecture yet.  We run a store with heavily customised product model and this plugin wouldn't run right for us anyway, we've added structured data as schema markup in our product page rather than the JSON object used here.  It looks very thorough, though, good work @torvista!

**Comment by torvista** [comment]
This functionality in the core?
Maybe. But only to add it to the list of ZC features **to attract new users**.

I think all the admin variables should be in a file, they never get changed once set up.

**Comment by neekfenwick** [comment]
> This functionality in the core? Maybe. But only to add it to the list of ZC features **to attract new users**.

Users who only ever use core ZC could probably benefit from this a great deal.  I'm not one.  But, the arguments made earlier in this Suggestion are valid.  A couple:

1. SEO does change, often more rapidly than major releases of ZC and even more rapidly than people updating their sites to those versions.  See @barco57 and @wiztechinc (referencing Facebook changes, and god knows I dropped a whole project because of that) making great points in this area.  I think it's a bad idea to tie a feature like SEO support to 'core' Zen Cart, when a plugin (or addon, or whatever they are called these days) could be updated independently of the main cart code base.
2. Some shops may have poorly set up products that would operate well enough as they are, but if they advertised that bad data as structured data to Google then their SEO would plummet faster than otherwise.  This SEO data is very powerful, I'd guess most of us shop owners have had a hellish week here and there trying to deal with SEO damage.  A nascent shop trying to get off the ground but hammered by awful data in structured data would kill potential ZC adherents before they got started.  To get around this you'd need a more complex solution that helps avoid these problems, and that leads back into the first point that the rules change and the code adhering to those rules needs to change fast.

I think it's pretty clear this is very cool stuff, but due to the nature of the industry it's aligning with, is not suited to be core ZC.

**Comment by torvista** [comment]
>due to the nature of the industry it's aligning with, is not suited to be core ZC.

And it's ugly/been constructed piecemeal/needs a rewrite.

Best that it is documented as an essential addition as part of setting up your shop, to pass the onus onto the owner to customise it correctly and understand that there will be ongoing changes.
